### PR TITLE
[ML] Only report complete writing_results progress after completion

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
We depend on the number of data frame rows in order to report progress
for the writing of results, the last phase of a job run. However, results
include other objects than just the data frame rows (e.g, progress, inference model, etc.).

The problem this commit fixes is that if we receive the last data frame row results
we'll report that progress is complete even though we still have more results to process
potentially. If the job gets stopped for any reason at this point, we will not be able
to restart the job properly as we'll think that the job was completed.

This commit addresses this by limiting the max progress we can report for the
writing_results phase before the results processor completes to 98.
At the end, when the process is done we set the progress to 100.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java [hunk 1]
```diff
@@ -107,11 +114,22 @@ public void process(AnalyticsProcess<AnalyticsResult> process) {
                 failure = "error parsing data frame analytics output: [" + e.getMessage() + "]";
             }
         } finally {
+            if (isCancelled == false) {
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java]: Should we take into account `failure != null`? It seems weird to me that we are "complete" even if we failed.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java]: Very good point. I also took the opportunity to improve failure reporting in that class in general.

**Review by benwtrent — APPROVED** [review]
:shipit:

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessorTests.java [hunk 2]
```diff
@@ -117,6 +119,28 @@ public void testProcess_GivenRowResults() {
...
+
+        AnalyticsResultProcessor resultProcessor = createResultProcessor();
+
+        resultProcessor.process(process);
+        resultProcessor.awaitForCompletion();
+
+        assertThat(resultProcessor.getFailure(), equalTo("error processing results; some failure"));
+
+        ArgumentCaptor<String> auditCaptor = ArgumentCaptor.forClass(String.class);
+        verify(auditor).error(eq(JOB_ID), auditCaptor.capture());
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessorTests.java]: Is it possible to replace `auditCaptor.capture()` with `containsString("Error processing results; some failure")`?

**Review by przemekwitek — APPROVED** [review]
LGTM

one minor comment in test code

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessorTests.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessorTests.java]: We need to call `capture()` on the captor for it to capture the argument to the mocked method.

**Pull request merged**